maro/simulator/scenarios/supply_chain/sku_dynamics_sampler.py

```diff
@@ -0,0 +1,294 @@
+from typing import Dict, Optional, Union
+
+from dateutil.parser import parse
+from tqdm import tqdm
+
+from maro.data_lib.supply_chain import (
+    DATE_INDEX_COLUMN_NAME, get_date_index, get_preprocessed_file_path, preprocess_file
+)
+
+from .units.seller import SellerDemandInterface
+
+if typing.TYPE_CHECKING:
+    from maro.simulator.scenarios.supply_chain.world import World
+
+
+DynamicsInfoItem = namedtuple("DynamicsInfoItem", ("column_name", "type_name", "default_value"))
+
+
+class SkuDynamicsSampler(metaclass=ABCMeta):
+    """Sampler to read SKU dynamics info from preprocessed data files, one store one file.
+    The preprocessed file has a `DateIndex` column which represents its date's index, starting from 01/01/1970.
+
+    NOTE:
+        This sampler need to configure the start time that to be treated as tick 0 in world.settings, or
+        it will use first row as start time.
+
+    Args:
+        configs (dict): Configuration of the facility it belongs to, it should contains following keys.
+            . "file_path", the path to the data file.
+            . "sku_column", column name contains sku name, must be in the facility's SKUs, or will be ignored.
+            . "price_column", column name that will be treated as price.
+            . "demand_column", column name that will be treated as demand from the end customers (sales).
+            . "datetime_column", column name that contains datetime, NOTE: we will parse it that ignore the time zone.
+        world (World): Current world this facility belongs to.
+    """
+    def __init__(self, configs: dict, world: World) -> None:
+        self._configs: dict = configs
+        self._world: World = world
+
+        self._file_path = configs["file_path"]
+        self._preprocessed_file_path = get_preprocessed_file_path(configs["file_path"])
+
+        self._start_date_index: Optional[int] = None
+        start_date_time = self._world.configs.settings["start_date_time"]
+        if start_date_time is not None:
+            self._start_date_index = get_date_index(parse(start_date_time))
+
+        self._datetime_column_name = configs.get("datetime_column", "DT")
+        self._sku_column_name = configs.get("sku_column", "SKU")
+        self._info_dict: Dict[str, DynamicsInfoItem] = self._init_info_dict()
+
+        # TODO: build up SC datapipeline, and use the processed files by default.
+        if not os.path.exists(self._preprocessed_file_path):
+            print(f"Preprocessed file {self._preprocessed_file_path} does not exist. Start preprocessing now.")
+            preprocess_file(self._file_path, date_column_name=self._datetime_column_name)
+
+        # TODO: add end_tick / max_tick parameter to close the file handler in time.
+        self._cache = defaultdict(dict)
+
+    @abstractmethod
+    def _init_info_dict(self) -> Dict[str, DynamicsInfoItem]:
+        raise NotImplementedError
+
+
+class OneTimeSkuDynamicsSampler(SkuDynamicsSampler):
```

Comment:
`_init_info_dict` not implemented in `OneTimeSkuDynamicsSampler` and `StreamSkuDynamicsSampler`. If it is allowd to leave it as an empty method, do not define it as an abstract method. Instead, leave it as a empty but concrete method and let the subclasses overwrite it when needed.

It they are abstract classes, add `metaclass=ABCMeta`.